test/e2e/storage/testsuites/provisioning.go

```diff
@@ -73,132 +76,88 @@ func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
-		input.testCase.NodeName = driver.GetDriverInfo().Config.ClientNodeName
-	}
-
-	return resource, input
-}
-
-func (p *provisioningTestSuite) execTest(driver TestDriver, pattern testpatterns.TestPattern) {
-	Context(getTestNameStr(p, pattern), func() {
-		var (
-			resource     provisioningTestResource
-			input        provisioningTestInput
-			needsCleanup bool
-		)
-
-		BeforeEach(func() {
-			needsCleanup = false
-			// Skip unsupported tests to avoid unnecessary resource initialization
-			skipUnsupportedTest(p, driver, pattern)
-			needsCleanup = true
-
-			// Create test input
-			resource, input = createProvisioningTestInput(driver, pattern)
-		})
-
-		AfterEach(func() {
-			if needsCleanup {
-				resource.cleanupResource(driver, pattern)
-			}
-		})
-
-		// Ginkgo's "Global Shared Behaviors" require arguments for a shared function
-		// to be a single struct and to be passed as a pointer.
-		// Please see https://onsi.github.io/ginkgo/#global-shared-behaviors for details.
-		testProvisioning(&input)
-	})
-}
-
-type provisioningTestResource struct {
-	driver TestDriver
-
-	claimSize string
-	sc        *storage.StorageClass
-	pvc       *v1.PersistentVolumeClaim
-}
+func (p *provisioningTestSuite) defineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	var (
+		f             = framework.NewDefaultFramework("provisioning")
+		dInfo         = driver.GetDriverInfo()
+		config        *PerTestConfig
+		driverCleanup func()
+		testCase      *StorageClassTest
+		pvc           *v1.PersistentVolumeClaim
+		sc            *storage.StorageClass
+	)
+
+	init := func() {
+		// Check preconditions.
+		if pattern.VolType != testpatterns.DynamicPV {
+			framework.Skipf("Suite %q does not support %v", p.tsInfo.name, pattern.VolType)
+		}
+		dDriver, ok := driver.(DynamicPVTestDriver)
+		if !ok {
+			framework.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.VolType)
+		}
 
-var _ TestResource = &provisioningTestResource{}
-
-func (p *provisioningTestResource) setupResource(driver TestDriver, pattern testpatterns.TestPattern) {
-	// Setup provisioningTest resource
-	switch pattern.VolType {
-	case testpatterns.DynamicPV:
-		if dDriver, ok := driver.(DynamicPVTestDriver); ok {
-			p.sc = dDriver.GetDynamicProvisionStorageClass("")
-			if p.sc == nil {
-				framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", driver.GetDriverInfo().Name)
-			}
-			p.driver = driver
-			p.claimSize = dDriver.GetClaimSize()
-			p.pvc = getClaim(p.claimSize, driver.GetDriverInfo().Config.Framework.Namespace.Name)
-			p.pvc.Spec.StorageClassName = &p.sc.Name
-			framework.Logf("In creating storage class object and pvc object for driver - sc: %v, pvc: %v", p.sc, p.pvc)
+		// Now do the more expensive test initialization.
+		config, driverCleanup = driver.CreateDriver(f)
+		claimSize := dDriver.GetClaimSize()
+		sc = dDriver.GetDynamicProvisionStorageClass(config, "")
```

Comment:
at any point should we call the `driver.SkipUnsupportedTest` here?

Reply:
It was already called, see https://github.com/kubernetes/kubernetes/blob/4410815d24a4a25899bb40cb9744b0ba9a12bd5e/test/e2e/storage/testsuites/base.go#L96-L123

Reply:
That test runs before the testsuite specific code.